bulwark/checks.py

```diff
@@ -408,6 +408,35 @@ def is_same_as(df, df_to_compare, **kwargs):
     return df
 
 
+def matches_regex(df, pattern, columns=None, **kwargs):
+    """Asserts that the values in `df`'s `columns` match `pattern`.
+
+    Args:
+        df (pd.DataFrame): Any pd.DataFrame.
+        pattern (str): Pattern to match against.
+        columns (list): A subset of columns to check for matching `pattern`.
+        **kwargs (dict): Keyword arguments passed through to pandas' ``Series.str.match``.
+
+    Returns:
+        Original `df`.
+
+    """
+    columns = columns if columns is not None else df.columns
+
+    non_matches_df = pd.DataFrame()
+
+    for col in columns:
+        matches = df[col].str.match(pattern, **kwargs).fillna(False)
+        if not matches.all():
+            non_matches_df[col] = ~matches
```

Comment:
This won't behave as expected in cases where there are multiple columns to check, and the bad values are in different rows, since index isn't preserved. For example:

```
df = pd.DataFrame({'A': ['a1', 'ab', 'ac'], 'B': ['bd', 'be', 'b1']})
pattern = r'a[a-z]'
ck.matches_regex(df, pattern, columns=['A', 'B'])
```

The expected result would be `AssertionError: [(0, 'A') (2, 'B')]`, but the result will be `AssertionError: [(0, 'A') (0, 'B') (1, 'B') (2, 'B')]`